### DISCUSSION TITLE
Automating Database Upgrades - Any progress?

### DISCUSSION THREAD

**Discussion by apmuthu** [timschofield/webERP#167] [Feature Requests]
[Automating Database Upgrades](http://www.weberp.org/wiki/AutomateDatabaseUpgrades) whither?

The config table parameter **VersionNumber** provides the currently installed version.
There is no parameter ** DatabaseVersion** yet.

Must we still carry it out manually?

Why does the `sql/mysql/upgrade4.14.1-4.14.2.sql` have in it's contents, the last line:
````
UPDATE config SET confvalue='4.15' WHERE confname='VersionNumber';
````
instead of:
````
UPDATE config SET confvalue='4.14.2' WHERE confname='VersionNumber';
````
???

Even inserting the following at [Line 202](https://github.com/timschofield/webERP-svn/blob/master/UpgradeDatabase.php#L202) in `UpgradeDatabase.php` just hangs when a fresh login is attempted:
````
				case '4.14.1':
					$SQLScripts[] = './sql/mysql/upgrade4.14.1-4.14.2.sql';
````

**Comment by timschofield** [comment]
I have long advocated a system where we have sequential numbered files containing atomic upgrades to the database. So if you write a new feature, then the DB upgrades for that functionality would be in one self contained file. When that file is processed it automatically updates the DB version number in the config table to be that number. When an admin user logs in, the highest number file in the upgrades directory is compared to the number in the config table, and if there are unprocessed files it processes them. It is simple and neat, and avoids all these problems you have been finding.

Tim

**Reply by gggeek** [reply]
I'd rather go for a slightly more complex table, where each change is stored in a dedicated line once executed. The system has a known list of directories where to scan for upgrade files, which can be either .php or .sql.
The principles can be seen in python migration, doctrine migrations, or in the kaliop-ezmigrationbundle (disclaimer: i wrote the latter).

**Comment by gggeek** [comment]
> The config table parameter VersionNumber provides the currently installed version.
> There is no parameter **DatabaseVersion** yet.

I looked at current code, and the situation seems to be that there is a `DBUpdateNumber` in the config, besides `VersionNumber`.

The weird thing is that:
- there is a integer-sequenced list of .php update files, which generally do change DB structure and/or data, then set the `DBUpdateNumber` to their integer number
- there is a list of sql files, with names corresponding to the webERP versions (from-to)
- there are 2 "upgrade the system" files: `UpgradeDatabase.php` and `Z_UpgradeDatabase.php`. Both are in use
- the code triggering the user being redirected to either of those is found both in `session.php` and `GetConfig.php`
- UpgradeDatabase.php does have a hardcoded list of the upgrade sql files, instead of just scanning for them

In short: things are messy and could definitely do with some refactoring and simplification.

**Reply by timschofield** [reply]
There are 2 systems, one uses UpgradeDatabase.php (for up to v4.15.2) and one that uses Z_UpgradeDatabase.php (for > 4.15.2).
We have no control over where people are upgrading from so both systems _must_ be left as is.

Thanks
Tim

**Reply by gggeek** [reply]
FWIW, I suggested in another thread that, post v5.0, we should drop support for upgrading from any version prior to 5.0. This means that any upgrade 3.x -> 5.0.1+ or 4.x -> 5.0.1+ would need to be done in 2 steps, by first installing v5.0 and running the upgrades in it, then installing the final version and running further upgrades. Imho the price to pay in user experience is well worth the advantage that we would get from dropping many upgrade files and being able to refactor the upgrade system...

**Comment by timschofield** [comment]
The user experience is far more important to me than having a few extra files in webERP.

Tim

**Comment by gggeek** [comment]
> The user experience is far more important to me than having a few extra files in webERP.

Once more, one of the issues we will have to agree to disagree about. My opinion is that the current codebase suffers heavily from cruft accumulated over the years (*), and that bringing it to a more manageable, lean and mean state is worthy of such inconveniences as in the proposed scenario.

* = case in point: see the monster PR I just sent to remove all usages of `$DbgMsg` in `DB_query` calls... there is way too much code running duplicate queries
